src/marker.rs

```diff
@@ -133,22 +112,46 @@ impl MarkerOptions {
 }
 
 pub struct Marker {
-    inner: crate::js::Marker,
-    latlng: crate::LngLat,
+    inner: js::Marker,
+    lnglat: LngLat,
+    handle: RefCell<Option<MarkerHandle>>,
 }
 
 impl Marker {
-    pub fn new(latlng: crate::LngLat, options: MarkerOptions) -> Marker {
-        let inner = crate::js::Marker::maker_new(options.build());
-        Marker { inner, latlng }
+    pub fn new(lnglat: LngLat) -> Rc<Marker> {
+        Self::with_listener(lnglat, MarkerOptions::default(), NoopListener {})
+    }
+
+    pub fn with_options(lnglat: LngLat, options: MarkerOptions) -> Rc<Marker> {
+        Self::with_listener(lnglat, options, NoopListener {})
+    }
+
+    pub fn with_listener<F>(lnglat: LngLat, options: MarkerOptions, f: F) -> Rc<Marker>
+    where
+        F: MarkerEventListener + 'static,
+    {
+        let marker = Rc::new(Marker {
+            inner: js::Marker::maker_new(options.build()),
+            lnglat,
+            handle: RefCell::new(None),
+        });
+
+        let handle = MarkerHandle::new(Rc::downgrade(&marker), f);
+        let inner = &marker.inner;
+        inner.Marker_on("dragstart".into(), &handle.on_dragstart);
+        inner.Marker_on("drag".into(), &handle.on_drag);
+        inner.Marker_on("dragend".into(), &handle.on_dragend);
+        marker.handle.borrow_mut().replace(handle);
+
+        marker
     }
 
-    pub fn add_to(&self, map: &crate::Map) {
-        self.inner.setLngLat(&self.latlng.inner);
+    pub(crate) fn add_to(&self, map: &Map) {
```

Comment:
makde `add_to` and `remove` private, because we want users to call `map.add_marker`/`map.remove_marker`.

Reply:
your right!